### PR TITLE
Release 0.7.1: complete the binding surface, the test suite, and the packaging

### PR DESCRIPTION
**Pull request by fametrano**
Brings `dev` up to `master` for the 0.7.1 release: 57 commits that widen the binding surface, replace the single smoke test with a real suite validated against external vectors, and rebuild packaging and CI around it.

## API surface

The wrappers now cover the modules that were previously reachable only through the raw extension:

- `keys`, `xonly` — key algebra, taproot tweaking, key ordering
- `dsa`, `ssa` — ECDSA signature forms, variable-length signing, tagged hashing
- `ecdh`, `recovery`, `ellswift`, `mult` — the remaining upstream modules
- `context` — the shared context, documented and tested under concurrent use
- `hashes`, `_scalar` — tagged hashing, and one helper that normalizes every scalar argument

Invalid arguments are reported consistently, return codes are checked, and what libsecp256k1 says through its own callbacks is surfaced rather than swallowed. MuSig2 is stated to belong to btclib; the ECDH hash function stays unexposed, with the reason recorded.

## Testing

`tests/test.py` is gone, replaced by tests exercised against **external** vectors only — official BIP340, RFC6979/DER, and btclib's own ECDSA vectors — plus property tests that assert invariants over derived inputs rather than chosen ones. Coverage is gated on a ratchet and measures only reachable code.

## Packaging and CI

- Vendored libsecp256k1 built with CMake on every platform
- Wheels for macOS arm64, Linux aarch64, Windows x64 and arm64; free-threaded interpreter tested, not merely built for
- `uv` manages the development environment; `uv.lock` is committed
- `ruff` replaces eight linters and formatters; `mypy --strict` type-checks the cffi boundary
- Least-privilege tokens, SHA-pinned actions, Dependabot for Actions and the submodule
- Tag-triggered PyPI release workflow with Trusted Publishing, able to rehearse itself on TestPyPI (a rehearsal is never tagged)
- `RELEASING.md` and `SECURITY.md` added; README and HISTORY brought current

## Note before merging

`pyproject.toml` still declares `version = "0.7.1rc1"`, set for the TestPyPI rehearsal. That needs to become `0.7.1` before the release tag is cut.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Finalize the 0.7.1 release by widening the libsecp256k1 binding surface, hardening argument validation and error reporting, rebuilding the CFFI/CMake-based extension and wheel packaging, and replacing the ad-hoc tests with a comprehensive, coverage-gated suite driven by external vectors and property tests.

New Features:
- Expose key algebra, x-only taproot operations, tagged hashing, ECDSA/Schnorr utilities, ECDH, recovery, ElligatorSwift, and raw MuSig2 access through dedicated binding modules.
- Add a shared, randomized libsecp256k1 context with per-thread callback reporting and document its concurrent, free-threaded-safe use.
- Provide versioned distribution metadata including PEP 639 licensing, `__version__`, and sdist/wheel contents suitable for Trusted Publishing with PEP 740 attestations.

Bug Fixes:
- Prevent invalid or out-of-range scalar, key, and signature inputs from crossing the CFFI boundary by normalizing and validating them consistently.
- Fix dynamic builds and sdists by moving callback stubs out of the vendored sources and ensuring CMake/CFFI subprocess failures are surfaced.
- Correct shared library discovery for dynamic wheels and ensure platform tags and deployment targets reflect the actual build environment.

Enhancements:
- Unify the vendored libsecp256k1 build on CMake across all platforms, avoid mutating the submodule, and support static and dynamic extension builds including cross-compiled Windows wheels.
- Refine ECDSA and Schnorr APIs to return booleans, support custom-length messages and auxiliary randomness, and expose normalization, malleability, and encoding conversion helpers.
- Strengthen type checking via strict mypy with dedicated stubs for the CFFI boundary, and introduce a scalar helper that standardizes all 32-byte numeric arguments.

Build:
- Switch to CMake as a declared build requirement for the vendored library, add a custom hatchling hook for CFFI modules, and tune wheel tags for static vs dynamic builds across CPython and PyPy, including free-threaded interpreters.
- Define sdist exclusions to ship pristine sources only and configure check-wheel-contents to accept ABI-mode layouts.
- Add uv/PEP 735 dependency groups and lockfile for reproducible local builds and benchmarking dependencies.

CI:
- Replace multiple workflows with a single `main` pipeline that runs linting, coverage, multi-platform wheel builds, dynamic builds, sdist builds, sdist install tests, distribution validation, and matrix testing across CPython and PyPy.
- Harden CI with least-privilege GITHUB_TOKEN, SHA-pinned actions, concurrency cancellation, and expanded platform matrices for macOS arm64, Linux aarch64, Windows x64 and arm64, and free-threaded Python.
- Introduce a `release` workflow that reuses the main pipeline, gates PyPI/TestPyPI uploads via environments and Trusted Publishing, and validates artifacts with twine and check-wheel-contents before publishing.

Deployment:
- Enable tag-triggered PyPI releases and manual TestPyPI rehearsals with clear release instructions and environment-based approval gates.
- Document and enforce the release versioning scheme and yanking policy for broken releases.

Documentation:
- Rewrite and expand the README to cover module coverage, versioning, MuSig2 rationale, thread safety, vendored-library decisions, build/test/dev workflow, and platform support.
- Add `RELEASING.md` detailing Trusted Publishing, rehearsals on TestPyPI, and one-time index setup, plus `SECURITY.md` describing the vulnerability-reporting process and the boundary with upstream libsecp256k1.
- Update HISTORY.md with a comprehensive 0.7.1 changelog and align metadata URLs with the current documentation and changelog locations.
- Document build and benchmark scripts in `scripts/README.md` and clarify their non-API status.

Tests:
- Replace the legacy single test script with a structured test suite covering BIP340 official vectors, RFC6979 and btclib ECDSA vectors, DER edge cases, libsecp256k1 module behaviour (ECDH, recovery, ElligatorSwift, MuSig2), core signing and input validation, extension loading, callbacks, concurrency, key algebra invariants, and tagged hashing.
- Introduce property-style tests that sweep derived inputs via SHA256 chains to exercise scalar/point algebra, ECDH/XDH symmetry, taproot tweaking, key ordering, and rare cases such as leading-zero x coordinates and short DER signatures.
- Configure pytest, coverage (branch mode with 100% fail-under ratchet and unreachable RuntimeError paths excluded), and cibuildwheel test hooks to ensure every wheel and interpreter combination is exercised.

Chores:
- Replace the linting/formatting stack (black, isort, flake8, bandit, pyupgrade, etc.) with ruff and update the pre-commit configuration, including mypy mirror settings and markdownlint updates.
- Add Dependabot configuration for GitHub Actions and the secp256k1 submodule and pin pre-commit hook revisions for regular updates.
- Drop deprecated Python versions (3.7, 3.8), add support for newer ones (3.13, 3.14 and free-threaded 3.14t), and adjust cibuildwheel and CI matrices accordingly.

### CODE REVIEW

File: README.md
Comment (giacomocaironi) [inline README.md:1]: I feel it goes into much more detailed than is needed in a README. Maybe create docs or point to a different file that explains what is actually validated by this library

File: SECURITY.md
Comment (giacomocaironi) [inline SECURITY.md:1]: Everything correct but way too verbose

File: pyproject.toml [hunk 1]
```diff
@@ -1,127 +1,345 @@
 [project]
 name="btclib_libsecp256k1"
-version="0.4.0"
+version="0.7.1rc1"
```
Comment (giacomocaironi) [inline pyproject.toml]: Correct but does not follow version convention stated elsewhere

File: pyproject.toml [hunk 2]
```diff
@@ -1,127 +1,345 @@
...
+    # !stubs/_btclib_libsecp256k1.pyi negation there and drops the stub
+    # the strict mypy gate needs from the sdist
+    "/_btclib_libsecp256k1.*",
+    "**/.libs",
+    "**/.deps",
+    "**/*.o",
+    "**/*.lo",
+    "**/*.la",
+    "**/*~",
+    "secp256k1/src/btclib_default_callbacks.c",
```
Comment (giacomocaironi) [inline pyproject.toml:104]: definetely can be a regex

File: pyproject.toml
Comment (giacomocaironi) [inline pyproject.toml:1]: Improved documentation and rationale behind certain choices is certainly useful, but I feel this is way, way too much

File: .python-version
Comment (giacomocaironi) [inline .python-version:1]: This file shouldn't be commited, is tied to the user

File: .secrets.baseline
Comment (giacomocaironi) [inline .secrets.baseline:1]: I don't believe there isn't a way to create a file with only the custom rules for our library instead of pulling this super big default file

File: .gitignore [hunk 3]
```diff
@@ -81,8 +80,8 @@ target/
 # Jupyter Notebook
 .ipynb_checkpoints
 
-# pyenv
-.python-version
+# pyenv / uv: .python-version is tracked, it pins the interpreter the
```
Comment (giacomocaironi) [inline .gitignore]: False, I may decide to use pyenv and use a custom virtualenv. This file needs to be excluded

File: scripts/README.md
Comment (giacomocaironi) [inline scripts/README.md:1]: Good idea, too verbose

File: scripts/cffi_build.py [hunk 4]
```diff
@@ -6,129 +6,200 @@
...
-                        artifacts.append(build_dir / file.name)
-                        found = True
-                    if not found:
-                        raise RuntimeError(f"no shared object found for library: {lib}")
-
-        return ffi, artifacts
+            return ffi, self.compile_static_unix(ffi, build_dir)
+        return ffi, self.emit_dynamic(ffi, build_dir)
+
+    def compile_static_msvc(
```
Comment (giacomocaironi) [inline scripts/cffi_build.py:119]: Wow I din't think it was this simple on windows. At this point the manual code for unix is really needed. I wrote it a long time ago

File: scripts/cffi_build.py [hunk 5]
```diff
@@ -138,86 +209,179 @@ def __init__(self):
...
 
     def clean(self) -> None:
-        subprocess.call(["git", "reset", "--hard"], cwd=self.wd)  # nosec B603 B607
-        subprocess.call(["git", "clean", "-fxd"], cwd=self.wd)  # nosec B603 B607
-        clean_libs = False
-        for libs_dir in self.library_dirs:
-            if libs_dir.exists():
-                clean_libs = True
-        if clean_libs:
-            subprocess.call(["make", "clean"], cwd=self.wd)  # nosec B603 B607
```
Comment (giacomocaironi) [inline scripts/cffi_build.py:158]: Why did you ditch make own cleanup?

**Review by giacomocaironi — DISMISSED** [review]
I approve the changes to the repo to bring it up to standards, support the latest python interpreters and libsecp256k1 versions. 
However this PR is made up of two main components, the aforementioned ones and the extension of the repo scope, with the addition of a lot more functions and the related tests. These two things are separate, and need to be addressed in two sequential PRs.

**Review by giacomocaironi — APPROVED** [review]
